File: src/lib/events.ts
/**
 * Typed Rust→Frontend event subscriptions (spec 02 §7).
 *
 * Event names mirror the constants in `src-tauri/src/events.rs` exactly. Each
 * `onX` helper wraps Tauri's `listen<T>` with the correct payload type and
 * returns the `UnlistenFn`. Stores are the only place that reconcile these
 * payloads into UI state (AGENTS.md §5); M4 wires them up.
 */

import {
  listen,
  type EventCallback,
  type UnlistenFn,
} from "@tauri-apps/api/event";
import type {
  AppSettings,
  ForwardConfig,
  ForwardStatus,
  LogEntry,
  TunnelGroup,
  TunnelStats,
  UpdateStatus,
} from "./types";

export const EVENTS = {
  tunnelStatus: "tunnel://status",
  tunnelStats: "tunnel://stats",
  logLine: "log://line",
  logCleared: "log://cleared",
  forwardsChanged: "forwards://changed",
  groupsChanged: "groups://changed",
  settingsChanged: "settings://changed",
  updateStatus: "update://status",
  updateProgress: "update://progress",
  windowFocus: "window://focus",
  /** Emitted to the `tray_popover` webview on every open (re-hydrate trigger). */
  trayOpened: "tray://opened",
} as const;

// --- Payload types (mirror the Rust emit structs, spec 02 §7) ---

export interface TunnelStatusEvent {
  id: string;
  status: ForwardStatus;
  lastError: string | null;
}

export interface TunnelStatsEvent {
  id: string;
  stats: TunnelStats;
}

export interface UpdateProgressEvent {
  downloaded: number;
  total: number | null;
}

// --- Typed subscription helpers ---

export const onTunnelStatus = (
  cb: EventCallback<TunnelStatusEvent>,
): Promise<UnlistenFn> => listen(EVENTS.tunnelStatus, cb);

export const onTunnelStats = (
  cb: EventCallback<TunnelStatsEvent>,
): Promise<UnlistenFn> => listen(EVENTS.tunnelStats, cb);

export const onLogLine = (cb: EventCallback<LogEntry>): Promise<UnlistenFn> =>
  listen(EVENTS.logLine, cb);

export const onLogCleared = (cb: EventCallback<null>): Promise<UnlistenFn> =>
  listen(EVENTS.logCleared, cb);

export const onForwardsChanged = (
  cb: EventCallback<ForwardConfig[]>,
): Promise<UnlistenFn> => listen(EVENTS.forwardsChanged, cb);

export const onGroupsChanged = (
  cb: EventCallback<TunnelGroup[]>,
): Promise<UnlistenFn> => listen(EVENTS.groupsChanged, cb);

export const onSettingsChanged = (
  cb: EventCallback<AppSettings>,
): Promise<UnlistenFn> => listen(EVENTS.settingsChanged, cb);

export const onUpdateStatus = (
  cb: EventCallback<UpdateStatus>,
): Promise<UnlistenFn> => listen(EVENTS.updateStatus, cb);

export const onUpdateProgress = (
  cb: EventCallback<UpdateProgressEvent>,
): Promise<UnlistenFn> => listen(EVENTS.updateProgress, cb);

export const onWindowFocus = (cb: EventCallback<null>): Promise<UnlistenFn> =>
  listen(EVENTS.windowFocus, cb);

export const onTrayOpened = (cb: EventCallback<null>): Promise<UnlistenFn> =>
  listen(EVENTS.trayOpened, cb);

File: src-tauri/src/events.rs
//! Event-name constants for Rust→Frontend events (spec 02 §7).
//!
//! Emitted via `AppHandle::emit`; the frontend subscribes to these exact
//! strings in `src/lib/events.ts`. Never emit an ad-hoc name not listed here
//! (AGENTS.md §1). Payload structs are added alongside the emitting subsystem
//! in later milestones; the wire shape for each is documented per constant.

/// `{ id: String, status: ForwardStatus, lastError: Option<String> }` — a
/// tunnel status transition.
pub const TUNNEL_STATUS: &str = "tunnel://status";

/// `{ id: String, stats: TunnelStats }` — stats update on the single 3s
/// sampler tick.
pub const TUNNEL_STATS: &str = "tunnel://stats";

/// `LogEntry` — a new log line was appended to the ring buffer.
pub const LOG_LINE: &str = "log://line";

/// `()` — the log buffer was cleared.
pub const LOG_CLEARED: &str = "log://cleared";

/// `Vec<ForwardConfig>` — the config list mutated (CRUD/reorder/migration).
pub const FORWARDS_CHANGED: &str = "forwards://changed";

/// `Vec<TunnelGroup>` — groups mutated.
pub const GROUPS_CHANGED: &str = "groups://changed";

/// `AppSettings` — settings changed (e.g. theme applied elsewhere).
pub const SETTINGS_CHANGED: &str = "settings://changed";

/// `UpdateStatus` — update availability changed.
pub const UPDATE_STATUS: &str = "update://status";

/// `{ downloaded: u64, total: Option<u64> }` — update download progress.
pub const UPDATE_PROGRESS: &str = "update://progress";

/// `()` — the window was re-shown (e.g. via single-instance); the frontend may
/// refresh.
pub const WINDOW_FOCUS: &str = "window://focus";

/// `()` — the tray popover was opened (tray left-click). Emitted **only** to the
/// `tray_popover` window so its panel UI rehydrates fresh tunnel/group/settings
/// state on every open (the popover webview may have been idle while hidden).
pub const POPOVER_OPENED: &str = "tray://opened";
